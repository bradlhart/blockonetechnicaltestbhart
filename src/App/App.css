.rootSegment {
    margin: 20px !important;
}
