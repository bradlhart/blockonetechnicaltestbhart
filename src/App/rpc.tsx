import { JsonRpc } from 'eosjs';

const rpc = new JsonRpc('https://api.eosnewyork.io');

export const rpcGetInfo = () => rpc.get_info();

export const rpcGetBlock = (block_num: number) => rpc.get_block(block_num);
