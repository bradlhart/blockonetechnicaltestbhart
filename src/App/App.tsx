import Promise from 'bluebird';
import React, { useState, useEffect } from 'react';
import { JsonRpc, RpcError } from 'eosjs';
import './App.css';

import { BlockSegment } from './BlockSegment/BlockSegment';

const rpc = new JsonRpc('https://api.eosnewyork.io');

const App: React.FC = () => {
  const [loading, setLoading] = useState(false);
  const [recentBlockSegments, setRecentBlockSegments] = useState([]);

  const loadBlockInformation = async () => {
    setLoading(true);
    setRecentBlockSegments([]);
    const chainInfo = await rpc.get_info();
    const recentBlockNum = chainInfo.last_irreversible_block_num;
    const rowCount = 10;
    const blockNumbers = [];
    for (let i = recentBlockNum; i > recentBlockNum - rowCount; i--) blockNumbers.push(i);
    const blockSegments = await Promise.mapSeries(blockNumbers, (blockNum: number) =>
      createBlockSegmentComponent(blockNum)
    );
    setRecentBlockSegments(blockSegments);
    setLoading(false);
  };

  const createBlockSegmentComponent = blockNum => {
    return new Promise(resolve => {
      rpc
        .get_block(blockNum)
        .then(blockInfo => {
          return resolve(<BlockSegment className="ui segment" key={blockInfo.block_num} blockInfo={blockInfo} />);
        })
        .catch(() => {
          return resolve(<div className="ui segment">Error Retrieving Block Number {blockNum}</div>);
        });
    });
  };

  return (
    <div className="ui container">
      <div className="ui segment" style={{ margin: '20px' }}>
        <div className="ui two column grid">
          <div className="fourteen wide column">
            <h2 className="ui header">Block.one Web Application Developer Technical Test</h2>
          </div>
          <div className="two wide column">
            <div className={'ui button ' + (loading ? 'loading disabled' : undefined)} onClick={loadBlockInformation}>
              LOAD
            </div>
          </div>
        </div>

        <div className="ui segments">
          <div className="ui segment">
            <div className="ui three column divided grid" style={{ fontWeight: 'bold' }}>
              <div className="ten wide column">Hash</div>
              <div className="three wide column">Timestamp</div>
              <div className="three wide column">Actions</div>
            </div>
          </div>
          {loading ? <div className="ui segment center aligned">Fetching Data...</div> : recentBlockSegments}
        </div>
      </div>
    </div>
  );
};

export default App;
