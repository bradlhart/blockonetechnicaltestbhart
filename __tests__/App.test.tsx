import React from 'react';
import { mount, shallow, ShallowWrapper } from "enzyme";

import App, { propagateBlockNumberList } from '../src/App/App';

jest.mock('../src/App/rpc', () => ({
  rpcGetBlock: () => {
    return {
      timestamp: "timestamp",
      producer: "producer",
      confirmed: 0,
      previous: "previous",
      transaction_mroot: "transaction_mroot",
      action_mroot: "action_mroot",
      schedule_version: 0,
      producer_signature: "producer_signature",
      id: "id",
      block_num: 0,
      ref_block_prefix: 0,
    };
  },
  rpcGetInfo: () => {
    return { last_irreversible_block_num: 9};
  }
}));

describe('App', () => {
  let wrapper: ShallowWrapper<any>;

  beforeEach(() => {
    wrapper = shallow(<App />);
  });

  test('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('loadBlockInformation', async () => {
    const container = mount(<App />);
    container.find('[data-testid="loadButton"]').simulate('click');


  });
});

describe('propagateBlockNumberList', () => {
  test('no lastBlockNum, full load', () => {
    const expected = [9, 8, 7, 6, 5, 4, 3, 2, 1, 0];
    const result = propagateBlockNumberList(undefined, 9, 10);
    expect(result).toEqual(expected);
  });

  test('lastBlockNum, partial load', () => {
    const expected = [9, 8, 7];
    const result = propagateBlockNumberList(6, 9, 10);
    expect(result).toEqual(expected);
  });
});
