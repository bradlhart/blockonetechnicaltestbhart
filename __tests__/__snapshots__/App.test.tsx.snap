// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders correctly 1`] = `
<div
  className="ui container"
>
  <div
    className="ui segment rootSegment"
  >
    <Header
      loadBlockInformation={[Function]}
      loading={false}
    />
    <List
      blockList={Array []}
      errorMsg=""
      loading={false}
    />
  </div>
</div>
`;
